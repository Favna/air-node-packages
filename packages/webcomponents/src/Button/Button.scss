@import '../styles/vars';

.button {
  @extend %fontNormal;
  background-color: $blue;
  border-radius: 0;
  box-shadow: none;
  float: right;
  height: 3.2rem;
  line-height: initial;
  margin-right: 1%;

  &Label {
    @extend %fontNormal;
    color: $white;
    font-size: 1rem;
    text-transform: none;
  }

  &Primary {
    border: 2px solid $blue;
  }

  &Secondary {
    background-color: $yellow;
    border: 2px solid $yellow;
    color: $textBlack;

    > span {
      color: $textBlack;
    }
  }

  &Outlined {
    background-color: $white;
    border: 2px solid $blue;
    color: $blue;

    > span {
      color: $blue;
    }

    &:hover {
      border-width: 2px;
    }

    &Secondary {
      background-color: $white;
      border: 2px solid $darkOrange;
      color: $textBlack;

      > span {
        color: $textBlack;
      }
    }
  }

  &Disabled {
    background-color: $lightGrey;
    border: transparent;
    color: $textBlack;

    > span {
      color: $textBlack;
    }
  }
}
