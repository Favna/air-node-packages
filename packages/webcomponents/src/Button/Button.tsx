/**
 * @class Button
 */

import MUIButton, { ButtonProps as MUIButtonProps } from '@material-ui/core/Button';
import css from './Button.scss';
import classnames from 'classnames';
import React, { FC, ReactNode, memo } from 'react';
import { customCss, dataQa } from '../constants';
import { makeStyles, Theme, createStyles } from '@material-ui/core';

export interface ButtonProps {
  /** The variant of the button */
  variant: 'text' | 'outlined' | 'contained' | undefined;
  /** The color type of the button */
  color: 'primary' | 'secondary' | 'default' | 'inherit' | undefined;
  /** The label to be displayed inside the button */
  label: ReactNode;
  /** Whether this button should be disabled */
  disabled?: boolean;
  /** Data-qa tag to apply to the search bar and input element */
  'data-qa'?: dataQa;
  /** Custom CSS classes to pass to the button */
  customclasses?: customCss;
  /** The action that should be triggered when clicking the button */
  onClick(): any;
}

const useStyles = makeStyles((theme: Theme) => createStyles({ buttonShadow: { boxShadow: theme.shadows[1], border: 'transparant' } }),
);

/**
 * Creates a button using pre-defined Rijkswatestaat styling
 *
 * ```jsx
 * import { Button } from '@rws-air/webcomponents'
 *
 * <Button label='button' onClick={() => undefined} variant='contained' color='primary' />
 * ```
 */
export const Button: FC<MUIButtonProps<'button', ButtonProps>> = props => {
  const classes = useStyles();

  return (
    <MUIButton
      data-qa={props['data-qa']}
      onClick={props.onClick}
      variant={props.variant}
      color={props.color}
      disabled={props.disabled}
      className={classnames(props.customclasses)}
      classes={{
        root: css.button,
        label: css.buttonLabel,
        disabled: css.buttonDisabled,
        containedPrimary: classnames(css.buttonPrimary, classes.buttonShadow),
        containedSecondary: classnames(css.buttonSecondary, classes.buttonShadow),
        outlined: classnames(css.buttonOutlined, classes.buttonShadow),
        outlinedPrimary: classnames(css.buttonOutlined, classes.buttonShadow),
        outlinedSecondary: classnames(css.buttonOutlinedSecondary, classes.buttonShadow),
      }}
      href={undefined}
    >
      {props.label}
    </MUIButton>
  );
};

export default memo(Button);