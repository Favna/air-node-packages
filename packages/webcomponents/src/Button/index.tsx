import { createStyles, makeStyles, Theme } from '@material-ui/core';
import MUIButton, { ButtonProps as MUIButtonProps } from '@material-ui/core/Button';
import clsx from 'clsx';
import React, { forwardRef, memo, ReactNode } from 'react';
import css from './Button.scss';

export interface ButtonProps {
  /** The variant of the button */
  variant: 'text' | 'outlined' | 'contained' | undefined;
  /** The color type of the button */
  color: 'primary' | 'secondary' | 'default' | 'inherit' | undefined;
  /** The label to be displayed inside the button */
  label: ReactNode;
  /** Whether this button should be disabled */
  disabled?: boolean;
  /** Data-qa tag to apply to the search bar and input element */
  'data-qa'?: string;
  /** Custom CSS classes to pass to the button */
  customclasses?: string | string[];
  /** Custom CSS classes to pass to the button label */
  customlabelclasses?: string | string[];
  /** The action that should be triggered when clicking the button */
  onClick(): unknown;
}

const useStyles = makeStyles((theme: Theme) =>
  createStyles({ buttonShadow: { boxShadow: theme.shadows[1], border: 'transparant' } })
);

/**
 * Constructs a button using pre-defined Rijkswaterstaat styling
 * @remark Ref is forwarded to the Material-UI button to allow for wrapping this button with a Tooltip
 * @param props Props to pass to the button
 * @example
 * ```jsx
 * <Button label='button' onClick={() => undefined} variant='contained' color='primary' />
 * ```
 */

export const Button = memo(
  forwardRef<HTMLButtonElement, MUIButtonProps<'button', ButtonProps>>((props, ref) => {
    const classes = useStyles();

    return (
      <MUIButton
        {...props}
        ref={ref}
        data-qa={props['data-qa']}
        onClick={props.onClick}
        variant={props.variant}
        color={props.color}
        disabled={props.disabled}
        className={clsx(props.customclasses)}
        classes={{
          root: css.button,
          label: clsx(css.buttonLabel, props.customlabelclasses),
          disabled: css.buttonDisabled,
          containedPrimary: clsx(css.buttonPrimary, classes.buttonShadow),
          containedSecondary: clsx(css.buttonSecondary, classes.buttonShadow),
          outlined: clsx(css.buttonOutlined, classes.buttonShadow),
          outlinedPrimary: clsx(css.buttonOutlined, classes.buttonShadow),
          outlinedSecondary: clsx(css.buttonOutlinedSecondary, classes.buttonShadow)
        }}
        href={undefined}
      >
        {props.label}
      </MUIButton>
    );
  })
);
