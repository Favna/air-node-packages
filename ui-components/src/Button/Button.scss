@import '../styles/vars';

.rwsButton {
  @include font-normal;
  background-color: $blue !important;
  border-radius: 0        !important;
  box-shadow: none        !important;
  float: right            !important;
  height: 3.2rem          !important;
  line-height: initial    !important;
  margin-right: 1%        !important;
}

.rwsButtonLabel {
  @include font-normal;
  color: $white           !important;
  font-size: 1rem         !important;
  font-weight: bold       !important;
  text-transform: none    !important;
}
