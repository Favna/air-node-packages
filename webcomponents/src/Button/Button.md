## Usage of the Button component:

**Example:**

```tsx
import React, { FC } from 'react'

import { Button } from '@rws-air/webcomponents'

const Example: FC = () => (
  <Button variant='contained' color='primary' label='button' onClick={() => console.log('void')} />
)
```

**Primary Button:**

```jsx
import { Button } from '@rws-air/webcomponents';

<Button variant='contained' color='primary' label='I am primary!' onClick={() => console.log('you clicked the basic button!')} />
```

**Disabled Button:**

```jsx
import { Button } from '@rws-air/webcomponents';

<Button variant='contained' color='primary' label='I am disabled!' disabled />
```

**Secondary Color button:**

```jsx
import { Button } from '@rws-air/webcomponents';

<Button
  variant='contained' color='secondary' label='I am secondary!'
  onClick={() => console.log('you clicked the secondary button!')} 
/>
```

**Outline Button:**

```jsx
import { Button } from '@rws-air/webcomponents';

<Button
  variant='outlined' color='primary' label='I am outlined!'
  onClick={() => console.log('you clicked the outlined button!')} 
/>
```