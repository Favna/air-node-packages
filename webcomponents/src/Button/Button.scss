@import '../styles/vars.scss';

.rwsButton {
  @include font-normal;
  background-color: $blue !important;
  border-radius: 0        !important;
  box-shadow: none        !important;
  float: right            !important;
  height: 3.2rem          !important;
  line-height: initial    !important;
  margin-right: 1%        !important;
}

.rwsButtonLabel {
  @include font-normal;
  color: $white           !important;
  font-size: 1rem         !important;
  font-weight: bold       !important;
  text-transform: none    !important;
}

button[class*="Secondary"] {
  background-color: $yellow !important;
}

button:disabled {
  background-color: $light-grey !important;
}

button[class*="outlined"] {
  background-color: $white !important;
  border: 2px solid $blue !important;
  color: $text-black !important;

  > span {
    color: $text-black !important;
  }
}