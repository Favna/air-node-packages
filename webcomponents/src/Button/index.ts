export { default as Button } from './Button';
export { RWSButtonProps as ButtonProps } from './Button';