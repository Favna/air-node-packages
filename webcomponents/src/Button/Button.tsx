/**
 * @class Button
 */

import React, { FC, ReactNode } from 'react'
import classnames from 'classnames';
import css from './Button.scss';
import { Button as MUIButton } from '@material-ui/core';
import { ButtonProps } from '@material-ui/core/Button';

import './Button.scss';

export type RWSButtonProps = {
    /** The variant of the button */
    variant: 'text' | 'outlined' | 'contained' | undefined;
    /** The color type of the button */
    color: 'primary' | 'secondary' | 'default' | 'inherit' | undefined;
    /** The label to be displayed inside the button */
    label: ReactNode;
    /** Whether this button should be disabled */
    disabled?: boolean;
    /** Any extra CSS classes to be passed to the button */
    customClasses?: string | string[];

    /** The action that should be triggered when clicking the button */
    onClick(): any;
};

export const Button: FC<ButtonProps<'button', RWSButtonProps>> = props => (
    <MUIButton
        onClick={props.onClick}
        variant={props.variant}
        color={props.color}
        disabled={props.disabled}
        className={classnames(props.customClasses)}
        classes={{ root: css.rwsButton, label: css.rwsButtonLabel }}
        href={undefined}
    >
        {props.label}
    </MUIButton>
)

export default Button;