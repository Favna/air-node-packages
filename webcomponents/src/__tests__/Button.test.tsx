import Button from '../Button/Button';
import { shallow, ShallowWrapper } from 'enzyme';
import React from 'react';

const mockCallback = jest.fn();
let button: ShallowWrapper;

beforeEach(() => button = shallow(<Button variant='contained' color='primary' label='testButton' onClick={mockCallback}/>));

test('button function called onClick', () => {
  button.simulate('click');
  expect(mockCallback).toHaveBeenCalled();
  expect(mockCallback).toHaveBeenCalledTimes(1);
});

test('button contains the correct translation', () => {
  expect(button.children().text()).toBe('testButton');
});