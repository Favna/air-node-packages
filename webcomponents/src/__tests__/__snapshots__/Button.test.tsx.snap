// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<WithStyles(ForwardRef(Button))
  className=""
  classes={
    Object {
      "containedPrimary": undefined,
      "containedSecondary": undefined,
      "disabled": undefined,
      "label": undefined,
      "outlined": undefined,
      "outlinedPrimary": undefined,
      "outlinedSecondary": undefined,
      "root": undefined,
    }
  }
  color="primary"
  onClick={[MockFunction]}
  variant="contained"
>
  testButton
</WithStyles(ForwardRef(Button))>
`;
